### PR TITLE
Feature/marketplace suggestions

### PR DESCRIPTION
**Pull request by haszari**
### Please re-review this PR…
This is a clean new PR, in woocommerce repo, for the code in https://github.com/woocommerce/woocommerce/pull/22712 (custom fork, PR now closed).

The feedback from that PR's reviews has been addressed. Please refer to the above closed PR for any review comments / discussion. 

When this PR is reviewed and approved, we'll use this branch/PR as the feature branch for Marketplace Suggestions. The follow up PRs will be merged into here once they pass review.

Apologies for the confusion – any questions/problems ping me @haszari and I'll sort it out. Thanks for your help!

cc @kloon @mikejolley @justinshreve @kovshenin @timmyc

### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Feature branch for Marketplace Suggestions. PRs relating to this feature will land here as they get approved.

Adds support for displaying subtle and appropriate in-product suggestion messaging for WooCommerce Marketplace extensions. These messages are driven by data from an API on WooCommerce.com.

### How to test the changes in this Pull Request:

* Checkout this branch and get it set up as the WooCommerce plugin in your development environment.
* Log in as a store administrator and view suggestions in the products admin screen – there are different suggestions displayed in the empty state and within the list of products.
* Test dismissing suggestions; dismissed suggestions should not show again.

This can also be tested on a [jurassic.ninja site](https://married-white-clawed.jurassic.ninja) - ping me if you want access.

### CODE REVIEW

**Comment by haszari** [comment]
The CI is failing on phpcs, on two existing files that I've touched.

How do we normally handle this – should these issues all be fixed as part of this PR?

The two files:

```
FILE: ...ome/travis/build/woocommerce/woocommerce/includes/admin/class-wc-admin.php
FOUND 29 ERRORS AFFECTING 24 LINES
....
FILE: ...is/build/woocommerce/woocommerce/includes/admin/helper/class-wc-helper.php
FOUND 98 ERRORS AFFECTING 84 LINES
```

File: includes/admin/helper/class-wc-helper.php
Comment (mikejolley) [inline includes/admin/helper/class-wc-helper.php]: Just a thought; store this transient as long as possible, and flush when a plugin is activated.

https://codex.wordpress.org/Plugin_API/Action_Reference/activated_plugin

This will mean cache is used more frequently.

File: includes/admin/marketplace-suggestions/class-wc-marketplace-suggestions.php
Comment (mikejolley) [inline includes/admin/marketplace-suggestions/class-wc-marketplace-suggestions.php]: Does json_decode need second argument to force assoc array vs object here? Not sure, just asking.

**Review by mikejolley — APPROVED** [review]
Looks good. I added one question, and one performance/cache suggestion.

R/e code standards, usually yes, but I'd leave them in this case to keep the branch/PRs clean.

**Review by kloon — APPROVED** [review]
LGTM, apart from the suggestions Mike mentioned.

Regarding the PHPCS fixes, you can do that in this PR in a separate commit, if it is too much and chances of introducing bugs then it is okay to skip it.

**Comment by timmyc** [comment]
@kloon and @mikejolley thanks for the reviews here 🙇 - should this be merged into `master` ( the current target of the PR ) or should we merge this to a new feature branch off of current `master`?

**Comment by kovshenin** [comment]
Can we get the other open PRs merged into this one for one final review?

**Comment by kloon** [comment]
We need to get all the PR's that forms part of this one merged into it and then have it re-reviewed. Unfortunately one of the confusing things around having multiple PR's open around the same functionality. We need to try and stick to a single PR for future features, just easier to review and test.

**Comment by timmyc** [comment]
Thanks for the clarification @kloon - so effectively this PR is the feature branch then correct?

**Comment by kloon** [comment]
> Thanks for the clarification @kloon - so effectively this PR is the feature branch then correct?
That is correct yes. There's a bunch of PR's that needs merging into this one still before we can do a propper review.

File: includes/admin/helper/class-wc-helper.php
Comment (kovshenin) [inline includes/admin/helper/class-wc-helper.php]: This will probably cause problems with the Helper.

We have no reliable way of determining when a new plugin is installed, or when one is deleted, and the Helper relies on this data always being accurate, so caching it, even for just a day, is going to cause confusion when users purchase a new extension, but the Helper doesn't see it for up to 24h.

The Helper will only call this method in the plugin (extension/subscription) management context, much like WordPress core. By context I mean the extension management screen in Woo core, the plugin list in wp-admin, as well as the plugin update checks in Cron. Caching this data for those routines has no real benefit in my opinion.

The problem we're trying to solve is that, what has previously been called in plugin management only, is now being called on many other non-plugin management screens in wp-admin, so caching sounds like a good solution. If we do cache it however, let's try and do this "internal" to the marketplace suggestions module to avoid any side effects on the Helper code.

**Review by kovshenin — CHANGES_REQUESTED** [review]
See comment about caching in `get_local_woo_plugins`.

File: includes/admin/helper/class-wc-helper.php
Comment (haszari) [reply includes/admin/helper/class-wc-helper.php]: Now fixed – I've reverted these changes, now we're using `active_plugins` option. 8dbda1b9c962a93be7d17ffa8ac55c2639f641e1

File: includes/admin/helper/class-wc-helper.php
Comment (haszari) [reply includes/admin/helper/class-wc-helper.php]: This code's gone now, using `active_plugins` instead, which is more conventional, and means we don't need to make changes in helper. 8dbda1b9c962a93be7d17ffa8ac55c2639f641e1

File: includes/admin/marketplace-suggestions/class-wc-marketplace-suggestions.php
Comment (haszari) [reply includes/admin/marketplace-suggestions/class-wc-marketplace-suggestions.php]: Correct – [the default is `bool $assoc = FALSE`](https://secure.php.net/manual/en/function.json-decode.php). This is fixed in the `WC_Queue` updater:

https://github.com/woocommerce/woocommerce/pull/22857/files#diff-7cb8cd3240aef8097b64fc2183b79fb1R66

**Comment by haszari** [comment]
This PR is ready to go from my / Team Alpha point of view.

- There are no follow up PRs to merge into this PR.
- This is blocked on @kovshenin review; however, the feedback has been addressed in the PR.

@timmyc @kloon What's the next step for merging this?

**Comment by timmyc** [comment]
@haszari excellent really happy to see the PR at this place! Looking at [circleci](https://travis-ci.org/woocommerce/woocommerce/jobs/499987111) it appears there are some phpcs fixes needed still but nothing major - but we will likely want to resolve those. 

I'll give this another 👀 from the JS side of things and a final 👍 - but thinking @kloon or @mikejolley will give it the final 👍 overall as that was the plan we discussed after getting all the related branches merged into this one.

File: assets/js/admin/marketplace-suggestions.js
Comment (timmyc) [inline assets/js/admin/marketplace-suggestions.js]: 💟thanks for this documentation right here, being able to easily locate where tracks are recorded is such a lovely thing.

File: assets/js/admin/marketplace-suggestions.js
Comment (timmyc) [inline assets/js/admin/marketplace-suggestions.js]: Minor but should this be on line 109? `} else {`

File: assets/js/admin/marketplace-suggestions.js
Comment (timmyc) [inline assets/js/admin/marketplace-suggestions.js]: Super minor but in this method and `renderSuggestionCTA` the var names of `left` and `right` which I'm guessing align with how they are ultimately displayed on the page seem a bit odd in the methods. I suppose something like `container` or `wrapper` might make more sense - kind of like how you call it a `row` in `renderTableBanner` below. Again super minor just made me pause when reading through the code.

File: assets/js/admin/marketplace-suggestions.js
Comment (timmyc) [inline assets/js/admin/marketplace-suggestions.js]: This needs to be translated.

**Review by timmyc — COMMENTED** [review]
Couple of pretty minor comments around the js logic, otherwise looking great to me. I really appreciate all the comments in the code throughout - made it easy to grok and follow along with.

File: assets/js/admin/marketplace-suggestions.js
Comment (haszari) [reply assets/js/admin/marketplace-suggestions.js]: Let's all promise to keep it updated when things change 🤝

File: assets/js/admin/marketplace-suggestions.js
Comment (haszari) [reply assets/js/admin/marketplace-suggestions.js]: The perils of lazy refactoring. Will look at fixing this as long as I'm confident it doesn't destabilise things :)

File: assets/js/admin/marketplace-suggestions.js
Comment (haszari) [reply assets/js/admin/marketplace-suggestions.js]: Good spotting, this is potentially a dumb default anyway. I'll wrap it and change to the more common `Learn More`.

Comment (haszari) [reply assets/js/admin/marketplace-suggestions.js]: This was one of those handy code review tips that revealed something less-than-ideal. I've now improved some related things:

- The default string is now `Learn More` – it's used for almost all content, so is a much better default
- The default applies for both button and link CTAs (was previously only button)
- The default CTA style is button, as the link is used much less often

**Comment by haszari** [comment]
Just to be clear @timmyc – this PR does not include a minified build of the JS changes (I couldn't get that working!).

I'd been hoping/assuming that there's a build process as part of the beta build/release (etc). If not, I'm happy to work with the experts to get my js changes built.

**Review by kovshenin — APPROVED** [review]

**Comment by mikejolley** [comment]
@haszari So I still have some problems with this PR before merging which I'll list below :)

1. The product edit screen section has no icon cmopared to the other sections

2. The behaviour when dismissing recommendations on the product list screen is super annoying. You dismiss, then when you come back you're immediately suggested something else. This feels way too intrusive. When you dismiss this should probably dismiss all suggestions for that area, at least for a period of time, otherwise everytime you refresh you're presenting something new which makes the dismissal appear broken.

3. The dismiss X itself needs a title or tooltip to label it as 'dismiss'.

4. Once you've dismissed everything on the product screen there is nothing that replaces the area, e.g. it could show 'No suggestions at present' or something with a link to the extension store. I see it dissapears after refresh only.

5. The suggestion on the product listing appears in a rather jarring way. It could probably slide out of something? I've accidently clicked it a few times because it pushes down the rest of the products suddenly without warning.

6. The scripts need to be minified. If you install `grunt` and run `grunt dev` there are a bunch of warnings that need to be fixed before this will work.

```
assets/js/admin/marketplace-suggestions.js
     33 |                    'slug': suggestionSlug,
                                                   ^ Extra comma. (it breaks older versions of IE)
     51 |            }
                      ^ Missing semicolon.
    101 |            linkoutButton.onclick = function( event ) {
                                                       ^ 'event' is defined but never used.
    105 |            }
                      ^ Missing semicolon.
    168 |                container.appendChild( renderDismissButton( slug ) )
                                                                             ^ Missing semicolon.
     29 |                ajaxurl,
                         ^ 'ajaxurl' is not defined.
    410 |})( jQuery, marketplace_suggestions, ajaxurl );
                                              ^ 'ajaxurl' is not defined.
    410 |})( jQuery, marketplace_suggestions, ajaxurl );
                     ^ 'marketplace_suggestions' is not defined.
      1 |/* global marketplace_suggestions ajaxurl */
         ^ 'marketplace_suggestions ajaxurl' is defined but never used.
   assets/js/frontend/add-to-cart-variation.js
    360 |                                    new_attr_select.find( 'option' ).each( function( index, el ) {
                                                                                                     ^ 'el' is defined but never used.
    360 |                                    new_attr_select.find( 'option' ).each( function( index, el ) {
                                                                                              ^ 'index' is defined but never used.
    360 |                                    new_attr_select.find( 'option' ).each( function( index, el ) {
                                                                                    ^ Functions declared within loops referencing an outer scoped variable may lead to confusing semantics. ($, attr_val, variation_active)
    386 |                    new_attr_select.find( 'option.attached.enabled' ).each( function( index, el ) {
                                                                                                      ^ 'el' is defined but never used.
    386 |                    new_attr_select.find( 'option.attached.enabled' ).each( function( index, el ) {
                                                                                               ^ 'index' is defined but never used.
```

Overall, the dismiss functionality feels broken. TBH, being able to dismiss each individual suggestion doesn't really seem beneficial, and causes real annoyances on the listing screens.

- Being able to dismiss on the blank slates and even the 'get more options' tab on the products screen doesn't really seem needed to me. These screens are only shown for a short period of time, and the product edit one is not visible unless you go to that tab. Those suggestions are super easy to ignore without feeling the need to dismiss them.
- On the flip side, being able to dismiss the more intrusive ones is essentual. That is, suggestions that appear inline in product lists, and any future suggestions that are visible on the page when it first loads. However, right now it's per suggestion. So when you have lots of suggestions it feels like the dismiss buttons are broken.

My suggestions:

1.  Only allow dismiss when a notice is visible on page load, on a screen which is used multiple times i.e. product list screen, not the blank slate or product edit page section.

2. Dismiss the area, not the suggestion. And/Or if dismissing the suggestion, have a cool down period before another suggestion is thrown back up in exactly the same place. e.g. a day.

cc @warrendholmes @avivapinchas @peterfabian

**Comment by kyleaparker** [comment]
>Only allow dismiss when a notice is visible on page load, on a screen which is used multiple times i.e. product list screen, not the blank slate or product edit page section.

I'd be ok with removing the dismiss on the products and orders empty state since they are temporary. On the product edit section I think it makes sense to individually dismiss the suggestions so we can keep showing new things. 

>Dismiss the area, not the suggestion. And/Or if dismissing the suggestion, have a cool down period before another suggestion is thrown back up in exactly the same place. e.g. a day.

I agree that we should wait before showing more and change the title of the product edit tab section to something like "No recommended extensions right now":

extension discovery i2 2019-03-04 11-37-07

**Comment by haszari** [comment]
Thanks for picking up those details @mikejolley ! 

> The dismiss X itself needs a title or tooltip to label it as 'dismiss'.
I've added a tooltip to clarify what it does – `Dismiss this suggestion`.

> Once you've dismissed everything on the product screen there is nothing that replaces the area
Good point. What I've done here is to slide out the heading if there are no suggestions.

I went with this approach as you can get in that state by dismissing the current suggestions, but you'll get more next time you're in there.

> The suggestion on the product listing appears in a rather jarring way.
Sliding out is problematic in a table; I couldn't get it to work. To make it less jarring it's now fading in.

> The scripts need to be minified. 
I fixed the jshint errors, thanks for bringing it to my attention! I manually ran the `uglify` step (`node_modules/grunt/bin/grunt uglify`). 

There are existing jshint issues in `assets/js/frontend/add-to-cart-variation.js`, I don't think it's appropriate for me to tackle those as part of this PR.

**Comment by kyleaparker** [comment]
>The product edit screen section has no icon cmopared to the other sections

This was intentional to make it clear the section is different - happy to revisit if it’s causing confusion

**Comment by mikejolley** [comment]
@kyleaparker No thats fine, just was not sure if it was intentionally left out.

**Comment by haszari** [comment]
> have a cool down period before another suggestion is thrown back up in exactly the same place. e.g. a day.

Have now implemented this – this little "snooze" interlude may make this stronger overall, thanks for the suggestion :)

File: assets/js/admin/marketplace-suggestions.js
Comment (timmyc) [inline assets/js/admin/marketplace-suggestions.js]: Testing this out with a backup from one of my "live" sites, and encountered an issue with this logic here. I have the Facebook extension installed which adds a column to the table, but other extensions do a similar thing, and users can also toggle columns on/off. So this is what the suggestion looks like to me:

colspan

We could lean on jquery a bit here to determine the right number of columns that are actively displayed:

`$( 'table.wp-list-table.posts thead th:not(.hidden)' ).length`

Which would give us the correct number of th's that are not currently hidden. I _think_ that would work. Along those lines too, but perhaps a bit more edge-case, is when users toggle column visibility, things would still get wonky, even if adjusting the colspan accordingly. We'd have to listen for those events and respond accordingly. Not sure if that is a critical problem or not, but something to consider.

**Review by timmyc — COMMENTED** [review]
Tested this out on a new data set / site with some extensions installed and hit one little bugger. Did verify that the dismissal of suggestions in the product listing worked as expected. Logged out, and logged back in and was not shown any suggestions still.

File: assets/js/admin/marketplace-suggestions.js
Comment (timmyc) [reply assets/js/admin/marketplace-suggestions.js]: Screen grab of toggling columns: https://drive.google.com/file/d/1IkapaDFhEFBGnB61D5kjXfpNv2Os9MNI/view?usp=sharing

Comment (haszari) [reply assets/js/admin/marketplace-suggestions.js]: Thanks for finding that – will look at implementing the jQuery based fix.

Comment (haszari) [reply assets/js/admin/marketplace-suggestions.js]: It's now calculating the `colspan` when the banner is initially displayed.

I've also added an event handler to catch when screen options change and recalculate it. There are probably some other cases we could catch – keeping it simple for now.

https://github.com/woocommerce/woocommerce/pull/22857/commits/f944da3eacaf72e381fb3a856dbe2e3f17612021

**Comment by nicomollet** [comment]
This is going too far.
I don't need suggestions, my shop managers don't need them.
Make this optional, thank you.

**Pull request merged**
